### PR TITLE
fixes #991 - Add ability to provide label to schema functions for filtered results

### LINKED ISSUE #991: Add ability to provide label to schema functions for filtered results
Currently when using e.g. apoc.schema.nodes() If i want to find out information about a specific label I have to post filter the results e.g. to find out constraints on 'Email' i would do:

CALL apoc.schema.nodes() yield label, properties
with * where label="Email" return properties

It would be great if we could provide the label directly to the procedure e.g.

CALL apoc.schema.nodes("Email") yield properties

### PR DESCRIPTION
**Pull request by AngeloBusato**
Fixes #991

Add ability to provide label to schema functions for filtered results

## Proposed Changes (Mandatory)

A brief list of proposed changes in order to fix the issue:

  - Now it can be possible to find all the nodes contraints/Indexes by label e.g (CALL apoc.schema.nodes(labelName) yield properties)

### CODE REVIEW

**Comment by jexp** [comment]
Can we change this to a consistent, optional config map that works for all schema/meta procedures.

My suggestion is to take it from apoc.meta.subGraph().

`{ labels:[], relationships:[], exclude:[], sample:number}`

Anything I miss? @sarmbruster @kvegter ?

**Comment by conker84** [comment]
Maybeis it better to split the `exclude` property?
`{ labels:[], relationships:[], excludeLabels:[], excludeRelationships:[], sample:number}`

**Comment by jexp** [comment]
Yes we could, makes it more explicit.
In general you shouldn't have an overlap between the two.

We could also work with a `"-"` prefix for exclusion, not sure which one is nicer for the user.

**Comment by jexp** [comment]
Perahaps also add a include/excluding of property-names.

**Comment by AngeloBusato** [comment]
@jexp PR Updated

**Comment by AngeloBusato** [comment]
@jexp PR updated

File: src/main/java/apoc/schema/SchemaConfig.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        return relationships;
+    }
+
+    public Set<String> getExcludeRelationships() {
+        return excludeRelationships;
+    }
+
+    public SchemaConfig(Map<String,Object> config) {
+        config = config != null ? config : Collections.emptyMap();
+        this.labels = (Set<String>) config.getOrDefault("labels", Collections.EMPTY_LIST);
```
Comment (jexp) [inline src/main/java/apoc/schema/SchemaConfig.java]: emptySet()

File: src/main/java/apoc/schema/Schemas.java [hunk 2]
```diff
@@ -44,15 +44,15 @@
     }
 
     @Procedure(value = "apoc.schema.nodes", mode = Mode.SCHEMA)
-    @Description("CALL apoc.schema.nodes() yield name, label, properties, status, type")
-    public Stream<IndexConstraintNodeInfo> nodes() throws IndexNotFoundKernelException {
-        return indexesAndConstraintsForNode();
+    @Description("CALL apoc.schema.nodes(config) yield name, label, properties, status, type")
```
Comment (jexp) [inline src/main/java/apoc/schema/Schemas.java]: config optional in descripton

Comment (jexp) [reply src/main/java/apoc/schema/Schemas.java]: also for the other procs

File: src/main/java/apoc/schema/Schemas.java [hunk 3]
```diff
@@ -296,12 +335,68 @@ private Boolean constraintsExistsForRelationship(String type, List<String> prope
...
+                        containsLabelName(excludes, relationship.getEndNode())) {
+                    relationshipIterator.remove();
+                }
+            }
+
+            return gr;
+        });
+    }
+
+    private boolean containsLabelName(Set<String> excludes, Node node) {
```
Comment (jexp) [inline src/main/java/apoc/schema/Schemas.java]: should probably go to util class

File: src/main/java/apoc/schema/Schemas.java [hunk 4]
```diff
@@ -269,15 +269,54 @@ private Boolean constraintsExistsForRelationship(String type, List<String> prope
...
         try ( Statement ignore = tx.acquireStatement() ) {
             TokenRead tokenRead = tx.tokenRead();
             TokenNameLookup tokens = new SilentTokenNameLookup(tokenRead);
 
             SchemaRead schemaRead = tx.schemaRead();
-            Iterable<IndexReference> indexesIterator = () -> schemaRead.indexesGetAll();
+            Iterable<IndexReference> indexesIterator;
+            Iterable<ConstraintDescriptor> constraintsIterator;
+
+            if (!includeLabels.isEmpty()) {
```
Comment (jexp) [inline src/main/java/apoc/schema/Schemas.java]: flip condition

**Comment by AngeloBusato** [comment]
@jexp PR Updated

**Pull request merged**